File: src/password_deriver/algorithms.py
import string
import argparse
import subprocess
import getpass
import hashlib
import random
import pathlib
import json
import sys

from . import base58


def derive_secret_key(master_password, domain, username, counter):
    salt = f"{domain}{username}{counter:x}".encode("utf-8")
    return hashlib.pbkdf2_hmac(
        "sha256", master_password, salt=salt, iterations=100_000, dklen=32
    )


def format_password_hex(secret_key):
    return base58.b58encode(secret_key[-12:]).decode("ascii") + "-"


def format_ethereum_private_key(secret_key):
    return secret_key.hex()


def derive_and_format(master_password, spec):
    domain = spec["domain"]
    username = spec["username"]
    method = spec["method"]
    if method == "v1-count3" or method == "v1-shorter-count3":
        counter = 3
    elif (
        method == "v1-count2"
        or method == "v1-shorter-count2"
        or method == "v1-with-bang-count2"
    ):
        counter = 2
    else:
        counter = 1
    secret_key = derive_secret_key(master_password, domain, username, counter)
    if method == "v1":
        return format_password_hex(secret_key)
    elif method == "v1-count2" or method == "v1-count3":
        return format_password_hex(secret_key)
    elif method == "v1-wo-tail":
        return format_password_hex(secret_key)[:-1]
    elif method == "v1-with-bang" or method == "v1-with-bang-count2":
        return format_password_hex(secret_key)[:-1] + "!"
    elif (
        method == "v1-shorter"
        or method == "v1-shorter-count2"
        or method == "v1-shorter-count3"
    ):
        return format_password_hex(secret_key)[:-3]
    elif method == "v1-shorter-with-dash":
        return format_password_hex(secret_key)[:-3] + "-"
    elif method == "ethereum":
        return format_ethereum_private_key(secret_key)
    elif method == "none":
        return ""
    else:
        raise ValueError(f"Unknown method: {method}.")

File: algorithms.go
package derk

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strconv"

	"golang.org/x/crypto/pbkdf2"
)

func deriveSecretKey(masterPassword string, domain string, username string, counter int) []byte {
	salt := []byte(fmt.Sprintf("%s%s%x", domain, username, counter))
	return pbkdf2.Key([]byte(masterPassword), salt, 100000, 32, sha256.New)
}

func formatBasic(secretKey []byte) string {
	return base58Encode(secretKey[len(secretKey)-12:])
}

func formatHex(secretKey []byte) string {
	return hex.EncodeToString(secretKey)
}

func DeriveAndFormat(masterPassword string, spec map[string]string) (string, error) {
	domain := spec["domain"]
	username := spec["username"]
	method := spec["method"]
	counter := 1
	counterStr, hasCounter := spec["counter"]
	if hasCounter {
		var err error
		counter, err = strconv.Atoi(counterStr)
		if err != nil {
			return "", err
		}
		if counter <= 0 {
			return "", fmt.Errorf("Counter has to be positive")
		}
	}

	secretKey := deriveSecretKey(masterPassword, domain, username, counter)

	switch method {
	case "v1":
		return formatBasic(secretKey) + "-", nil
	case "v1-wo-tail":
		return formatBasic(secretKey), nil
	case "v1-with-bang":
		return formatBasic(secretKey) + "!", nil
	case "v1-shorter":
		hx := formatBasic(secretKey)
		return hx[:len(hx)-2], nil
	case "v1-short":
		hx := formatBasic(secretKey)
		if len(hx) < 12 {
			return "", fmt.Errorf("Basic password is too short")
		}
		return hx[:12], nil
	case "ethereum":
		return formatHex(secretKey), nil
	default:
		return "", fmt.Errorf("Unknown method: %s", method)
	}
}
